Heads-up: 458e7b3 ("rename busy instance updater") missed the only production caller — `lib/llm/src/discovery/worker_monitor.rs:680` still calls `client.update_free_instances(...)`, which no longer exists. That's why every build job that compiles `dynamo-llm` is failing:

```
error[E0599]: no method named `update_free_instances` found for struct `dynamo_runtime::component::Client`
   --> lib/llm/src/discovery/worker_monitor.rs:680:36
    |
680 |     client.update_free_instances(&busy_instances);
    |            ^^^^^^^^^^^^^^^^^^^^^
```

(Tried to leave an inline suggestion but GitHub refuses since `worker_monitor.rs` isn't in this PR's diff — `path could not be resolved`.)

One-line fix in `lib/llm/src/discovery/worker_monitor.rs` around line 680:

```diff
                         if busy_instances != previous_busy_instances {
                             let total_workers = client.instance_ids().len();
-                            client.update_free_instances(&busy_instances);
+                            client.set_busy_instances(&busy_instances);
                             let free_workers = client.instance_ids_free().len();
```

(`rg "update_free_instances" lib/` confirms this is the only remaining caller now that the test sites in `client.rs` are already renamed.)

--- a/lib/llm/src/discovery/watcher.rs
+++ b/lib/llm/src/discovery/watcher.rs
@@ -804,7 +804,7 @@ impl ModelWatcher {
             // worker TTFT/ITL cleanup). The thresholds control busy detection behavior only.
             //
             // IMPORTANT: When KV routing is active, the monitor must use the KvRouter's Client
-            // so that busy-state updates (via update_free_instances) are visible to the
+            // so that busy-state updates (via set_busy_instances) are visible to the
             // PushRouter, which also uses the KvRouter's Client (see common.rs:258-263).
             // Using a different Client instance would cause the PushRouter to never see
             // busy workers, since each Client::new() creates independent ArcSwap state.

--- a/lib/llm/src/discovery/worker_monitor.rs
+++ b/lib/llm/src/discovery/worker_monitor.rs
@@ -677,7 +677,7 @@ impl WorkerLoadMonitor for KvWorkerMonitor {
                         // Only update if busy_instances has changed
                         if busy_instances != previous_busy_instances {
                             let total_workers = client.instance_ids().len();
-                            client.update_free_instances(&busy_instances);
+                            client.set_busy_instances(&busy_instances);
                             let free_workers = client.instance_ids_free().len();
                             tracing::debug!(
                                 busy_instances = ?busy_instances,

--- a/lib/runtime/src/component/client.rs
+++ b/lib/runtime/src/component/client.rs
@@ -234,6 +234,32 @@ impl Client {
             .collect()
     }
 
+    pub fn instance_count(&self) -> usize {
+        self.instance_source.borrow().len()
+    }
+
+    pub fn busy_instance_count(&self) -> usize {
+        self.instance_busy.load().len()
+    }
+
+    pub fn all_instances_busy(&self) -> bool {
+        let busy_ids = self.instance_busy.load();
+        if busy_ids.is_empty() {
+            return false;
+        }
+
+        let instances = self.instance_source.borrow();
+        let mut saw_instance = false;
+        for instance in instances.iter() {
+            saw_instance = true;
+            if !busy_ids.contains(&instance.id()) {
+                return false;
+            }
+        }
+
+        saw_instance
+    }
+
     /// Get a watcher for available instance IDs
     pub fn instance_avail_watcher(&self) -> tokio::sync::watch::Receiver<Vec<u64>> {
         self.instance_avail_rx.clone()

--- a/lib/runtime/src/component/client.rs
+++ b/lib/runtime/src/component/client.rs
@@ -234,14 +234,6 @@ impl Client {
             .collect()
     }
 
-    pub fn instance_count(&self) -> usize {
-        self.instance_source.borrow().len()
-    }
-
-    pub fn busy_instance_count(&self) -> usize {
-        self.instance_busy.load().len()
-    }
-
     pub fn all_instances_busy(&self) -> bool {
         let busy_ids = self.instance_busy.load();
         if busy_ids.is_empty() {

--- a/lib/runtime/src/pipeline/network/egress/push_router.rs
+++ b/lib/runtime/src/pipeline/network/egress/push_router.rs
@@ -448,7 +448,7 @@ where
     /// The rejection path is gated by `fault_detection_enabled` (true here);
     /// busy detection itself is driven by the monitor via `client.set_busy_instances(...)`.
     /// If no thresholds are configured on the monitor (or no monitor is provided),
-    /// `client.instance_ids_free()` returns all instances and the gate never rejects.
+    /// `client.all_instances_busy()` returns false and the gate never rejects.
     pub async fn from_client_with_monitor(
         client: Client,
         router_mode: RouterMode,
@@ -808,36 +808,33 @@ where
 
         // Check if all workers are busy (when fault detection is enabled).
         if self.fault_detection_enabled {
-            let free_instances = self.client.instance_ids_free();
+            let all_instances_busy = self.client.all_instances_busy();
             if tracing::enabled!(tracing::Level::DEBUG) {
                 tracing::debug!(
                     request_id = %request_id,
                     instance_id,
                     router_mode = ?self.router_mode,
-                    free_workers = free_instances.len(),
-                    total_workers = self.client.instance_ids().len(),
+                    busy_workers = self.client.busy_instance_count(),
+                    total_workers = self.client.instance_count(),
+                    all_instances_busy,
                     "checked worker busy state"
                 );
             }
-            if free_instances.is_empty() {
-                // Check if we actually have any instances at all
-                let all_instances = self.client.instance_ids();
-                if !all_instances.is_empty() {
-                    tracing::warn!(
-                        instance_id,
-                        total_workers = all_instances.len(),
-                        "Rejecting request: all workers are busy"
-                    );
-                    let cause = PipelineError::ServiceOverloaded(
-                        "All workers are busy, please retry later".to_string(),
-                    );
-                    return Err(DynamoError::builder()
-                        .error_type(ErrorType::ResourceExhausted)
-                        .message("All workers are busy, please retry later")
-                        .cause(cause)
-                        .build()
-                        .into());
-                }
+            if all_instances_busy {
+                tracing::warn!(
+                    instance_id,
+                    total_workers = self.client.instance_count(),
+                    "Rejecting request: all workers are busy"
+                );
+                let cause = PipelineError::ServiceOverloaded(
+                    "All workers are busy, please retry later".to_string(),
+                );
+                return Err(DynamoError::builder()
+                    .error_type(ErrorType::ResourceExhausted)
+                    .message("All workers are busy, please retry later")
+                    .cause(cause)
+                    .build()
+                    .into());
             }
         }
 

--- a/lib/runtime/src/pipeline/network/egress/push_router.rs
+++ b/lib/runtime/src/pipeline/network/egress/push_router.rs
@@ -814,8 +814,7 @@ where
                     request_id = %request_id,
                     instance_id,
                     router_mode = ?self.router_mode,
-                    busy_workers = self.client.busy_instance_count(),
-                    total_workers = self.client.instance_count(),
+                    total_workers = self.client.instance_source.borrow().len(),
                     all_instances_busy,
                     "checked worker busy state"
                 );
@@ -823,7 +822,7 @@ where
             if all_instances_busy {
                 tracing::warn!(
                     instance_id,
-                    total_workers = self.client.instance_count(),
+                    total_workers = self.client.instance_source.borrow().len(),
                     "Rejecting request: all workers are busy"
                 );
                 let cause = PipelineError::ServiceOverloaded(

